### PR TITLE
feat: request and reply decorators can not be a reference type

### PR DESCRIPTION
**Pull request by Eomm**
Implements https://github.com/fastify/fastify/issues/2598

This PR replaces the WARN message with an error when a reference type is used as a request or reply decorator, suggesting using a different API.

To use a reference type as a decorator without sharing values across requests, the user must use the `getter/setter` interface or the `hook` approach. Both methods are documented with examples.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/decorate.js [hunk 1]
```diff
@@ -56,9 +56,19 @@ function decorateConstructor (konstructor, name, fn, dependencies) {
   }
 }
 
-function checkReferenceType (name, fn) {
+function encapsulateReferenceType (name, fn) {
   if (typeof fn === 'object' && fn && !(typeof fn.getter === 'function' || typeof fn.setter === 'function')) {
-    FSTDEP006(name)
+    return encapsulatedDecorator(fn)
```
Comment (mcollina) [inline lib/decorate.js:61]: I would just throw in this case and not clone.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: lib/decorate.js (hunk 1 above)
Comment (Eomm) [reply lib/decorate.js:61]: So are we removing the warning to replace it with an error while suggesting the user use the getter/setter interface?

Comment (mcollina) [reply lib/decorate.js:61]: Yes.

Comment (Eomm) [reply lib/decorate.js]: Done

**Review by gurgunday — APPROVED** [review]
LGTM!

**Review by mcollina — APPROVED** [review]
code lgtm

Can you update the PR title and description?

**Review by metcoder95 — APPROVED** [review]

**Comment by melroy89** [comment]
Docs page forgotten to be updated: https://fastify.dev/docs/latest/Reference/Decorators/#decorators

> 
> // Decorate request with a 'user' property
> fastify.decorateRequest('user', '')

**Pull request merged**